commit 2c89c887a818c1671d1b53985f134f56fe1d2004
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Fri Oct 14 08:51:11 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Fri Oct 14 08:51:11 2022 -0400

    Enhance job run status (#1023)
    
    * WIP: update the job_run_status to parent process.
    
    * Added update_job_run_status to update the proper job run status.
    
    * code style fix.
    
    * fixed unittests.
    
    * rolled back the wait_for_complete logic.
---
 nvflare/private/fed/server/job_runner.py    |  7 ++++---
 nvflare/private/fed/server/server_engine.py | 28 ++++++++++++++++++++++++++--
 2 files changed, 30 insertions(+), 5 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 3251da15..928b8573 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -285,10 +285,10 @@ class JobRunner(FLComponent):
                     with self.lock:
                         job = self.running_jobs.get(job_id)
                         if job:
-                            execution_exception_run_processes = engine.execution_exception_run_processes
-                            if job_id in execution_exception_run_processes:
+                            exception_run_processes = engine.exception_run_processes
+                            if job_id in exception_run_processes:
                                 self.log_info(fl_ctx, f"Try to abort run ({job_id}) on clients.")
-                                run_process = execution_exception_run_processes[job_id]
+                                run_process = exception_run_processes[job_id]
 
                                 # stop client run
                                 client_sites = run_process.get(RunProcessKey.PARTICIPANTS)
@@ -301,6 +301,7 @@ class JobRunner(FLComponent):
                             fl_ctx.set_prop(FLContextKey.CURRENT_JOB_ID, job.job_id)
                             self.fire_event(EventType.JOB_COMPLETED, fl_ctx)
                             self.log_debug(fl_ctx, f"Finished running job:{job.job_id}")
+                    engine.remove_exception_process(job_id)
             time.sleep(1.0)
 
     def _save_workspace(self, fl_ctx: FLContext):
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 58463e52..4ea21fd6 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -95,7 +95,7 @@ class ServerEngine(ServerEngineInternalSpec):
         self.server = server
         self.args = args
         self.run_processes = {}
-        self.execution_exception_run_processes = {}
+        self.exception_run_processes = {}
         self.run_manager = None
         self.conf = None
         # TODO:: does this class need client manager?
@@ -241,9 +241,21 @@ class ServerEngine(ServerEngineInternalSpec):
                             targets=targets, topic=topic, request=request, timeout=timeout, fl_ctx=fl_ctx
                         )
                         conn.send(replies)
+                    elif command == ServerCommandNames.UPDATE_RUN_STATUS:
+                        execution_error = data.get("execution_error")
+                        if execution_error:
+                            with self.lock:
+                                run_process_info = self.run_processes.get(job_id)
+                                self.exception_run_processes[job_id] = run_process_info
+
             except BaseException as e:
                 self.logger.warning(f"Failed to process the child process command: {secure_format_exception(e)}")
 
+    def remove_exception_process(self, job_id):
+        with self.lock:
+            if job_id in self.exception_run_processes:
+                self.exception_run_processes.pop(job_id)
+
     def wait_for_complete(self, job_id):
         while True:
             try:
@@ -258,7 +270,7 @@ class ServerEngine(ServerEngineInternalSpec):
                         return_code = run_process_info[RunProcessKey.CHILD_PROCESS].poll()
                         # if process exit but with Execution exception
                         if return_code and return_code != 0:
-                            self.execution_exception_run_processes[job_id] = run_process_info
+                            self.exception_run_processes[job_id] = run_process_info
                 self.engine_info.status = MachineStatus.STOPPED
                 break
 
@@ -566,6 +578,18 @@ class ServerEngine(ServerEngineInternalSpec):
             return_data = self.parent_conn.recv()
             return return_data
 
+    def update_job_run_status(self):
+        with self.parent_conn_lock:
+            with self.new_context() as fl_ctx:
+                execution_error = fl_ctx.get_prop(FLContextKey.FATAL_SYSTEM_ERROR, False)
+                data = {
+                    ServerCommandKey.COMMAND: ServerCommandNames.UPDATE_RUN_STATUS,
+                    ServerCommandKey.DATA: {
+                        "execution_error": execution_error,
+                    },
+                }
+                self.parent_conn.send(data)
+
     def aux_send(self, targets: [], topic: str, request: Shareable, timeout: float, fl_ctx: FLContext) -> dict:
         # Send the aux messages through admin_server
         request.set_peer_props(fl_ctx.get_all_public_props())
